workspace.spec

```diff
@@ -572,6 +572,378 @@ module Workspace {
+			"CRediT:software"
+			"CRediT:supervision"
+			"CRediT:validation"
+			"CRediT:visualization"
+			"CRediT:writing-original-draft"
+			"CRediT:writing-review-editing"
+	*/
+	typedef structure {
+		string contributor_type;
+		string name;
+		string given_name;
+		string family_name;
+		string contributor_id;
+		list<Organization> affiliations;
+		list<string> contributor_roles;
+	} Contributor;
+
+	/*
+		Represents the credit metadata associated with a workspace object.
+
+		Required fields are:
+		- identifier
+		- resource_type
+		- titles (one or more required)
+		- contributors (one or more required)
+		- versioning information: if the resource does not have an explicit version number, one or more dates should be supplied, ideally the date of resource publication and the last update (if applicable).
+
+		string identifier - persistent unique identifier for the resource.
+			Should be in the format
+			<database name>:<identifier within database>
+			e.g.	DOI:10.46936/10.25585/60000745
+				CrossRef:IsManifestationOf
+				GO:0005456
+				HGNC:7470
+		string license (optional) - usage license for the resource.
+		string resource_type - the type of resource being represented.
+		string version (optional) - the version of the resource.
+		list<Contributor> contributors (optional) - a list of people and/or organizations who contributed to the resource.
+		list<EventDate> dates - a list of relevant lifecycle events for the resource.
+		list<FundingReference> funding (optional) - zero or more funding sources for the resource.
+		list<PermanentID> related_identifiers (optional) - zero or more PIDs related to the resource.
+		list<Title> titles - one or more titles for the resource.
+	*/
+	typedef structure {
+		string identifier;
+		string license;
+		string version;
+		string resource_type;
+		list<Contributor> contributors;
+		list<EventDate> dates;
+		list<FundingReference> funding;
+		list<PermanentID> related_identifiers;
+		list<Title> titles;
+	} CreditMetadata;
+
+	/* CreditMetadataContainer
+
+		Container for credit metadata; used when adding a new set of credit metadata
+		to a workspace object.
+
+		All fields are required.
+
+		string schema_version - version of the credit metadata schema used
+		CreditMetadata credit_metadata - the credit metadata itself
+	*/
+	typedef structure {
+		CreditMetadata credit_metadata;
+		string schema_version;
+	} CreditMetadataContainer;
+
+	/* CreditMetadataEntry
+
+		Container for an instance of credit metadata; returned by the workspace when
+		credit metadata is requested.
+
+		All fields will be populated.
+
+		username saved_by - the user who added this entry
+		epoch timestamp - timestamp for the addition of this credit metadata
+		string schema_version - version of the credit metadata schema used
+		CreditMetadata credit_metadata - the credit metadata itself
+	*/
+	typedef structure {
+		username saved_by;
+		epoch timestamp;
+		CreditMetadata credit_metadata;
+		string schema_version;
```

Comment:
I'm having qualms about having a schema version. Any backward incompatible change (or even forwards compatible change) should result in at workspace version bump (accessible programatically and cheaply via `ver()`). Is that not enough?

Reply:
I'd still want to have the version of the credit metadata schema easily available, though -- otherwise you would need to have a mapping of workspace version to credit metadata schema version.

Reply:
For adding/updating records, though, since the workspace API + data structures are going to be so closely linked to the CM schema, the CM schema version can be added automatically.

Reply:
I'm saying I don't think you need a schema version at all. When you say "the credit schema version is changing" what you're really saying is "the workspace API is changing". We don't have a schema version for provenance, say, and that's been fine for 9 years.

The way we manage that kind of thing would be
a) avoid backwards incompatible changes like the plague, and if that's impossible
b) have a toggle so the user can choose which version of the API data structures they want, with the understanding that older (default) versions may have less information.

Reply:
I def think we want to have transparent versioning of the credit metadata. This isn't really linked to workspace API updates, this is more linked to external changes in the publishing infrastructure, changes in the DataCite schema, etc. This is implemented in our workspace, but isn't really about us :)

Reply:
Resolved by in-person meeting.

Reply:
specifically, the credit schema will be checked into the workspace and the classes in the workspace will be kept in sync with that schema (with appropriate documentation so that new developers know this), thus removing a circular dependency (e.g. the credit engine depends on the workspace, which depends on a spec stored in the CE to define its class structure for credit data

Reply:
The rationale is that the schemas at 3rd party credit sites change fairly rapidly, and the credit schema here is based on those schemas. This does mean, however, that given multiple versions of the credit schema will be stored in the workspace at the same time, any API and internal classes dealing with credit must be able to represent any version of the schema.